# 목표
- 데이터 로딩
- 데이터 확인
- 데이터 색인(필터링)
- 데이터 연산

## 1) 데이터 로딩

In [5]:
import kagglehub
import pandas as pd
# Download latest version
path = kagglehub.dataset_download("melihkanbay/police")

print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/police


In [16]:
df = pd.read_csv(path+'/police.csv')

In [17]:
df.head()

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0  2005-01-02     01:55          NaN             M          1985.0   
1  2005-01-18     08:15          NaN             M          1965.0   
2  2005-01-23     23:15          NaN             M          1972.0   
3  2005-02-20     17:15          NaN             M          1986.0   
4  2005-03-14     10:00          NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

## 2) 데이터 확인

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           91741 non-null  object 
 1   stop_time           91741 non-null  object 
 2   county_name         0 non-null      float64
 3   driver_gender       86406 non-null  object 
 4   driver_age_raw      86414 non-null  float64
 5   driver_age          86120 non-null  float64
 6   driver_race         86408 non-null  object 
 7   violation_raw       86408 non-null  object 
 8   violation           86408 non-null  object 
 9   search_conducted    91741 non-null  bool   
 10  search_type         3196 non-null   object 
 11  stop_outcome        86408 non-null  object 
 12  is_arrested         86408 non-null  object 
 13  stop_duration       86408 non-null  object 
 14  drugs_related_stop  91741 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 9.3+

In [34]:
# 데이터가 없는 열 삭제하기
police = df.copy()
police = police.drop(labels="county_name", axis=1)

police.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           91741 non-null  object 
 1   stop_time           91741 non-null  object 
 2   driver_gender       86406 non-null  object 
 3   driver_age_raw      86414 non-null  float64
 4   driver_age          86120 non-null  float64
 5   driver_race         86408 non-null  object 
 6   violation_raw       86408 non-null  object 
 7   violation           86408 non-null  object 
 8   search_conducted    91741 non-null  bool   
 9   search_type         3196 non-null   object 
 10  stop_outcome        86408 non-null  object 
 11  is_arrested         86408 non-null  object 
 12  stop_duration       86408 non-null  object 
 13  drugs_related_stop  91741 non-null  bool   
dtypes: bool(2), float64(2), object(10)
memory usage: 8.6+ MB


## 3) 데이터 색인
- 2005년 1월 데이터만 보기
- 남성인 데이터만 보기
- 20세인 데이터만 보기

In [43]:
print("==== 2005년 1월인 데이터 ====")
display(police[police['stop_date'].str.startswith("2005-01")])

print("\n\n==== 남성인 데이터 ====")
display(police[police['driver_gender'] == "M"])

print("\n\n==== 20세 데이터 ====")
display(police[police['driver_age'] == 20])

==== 2005년 1월인 데이터 ====


stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0  2005-01-02     01:55             M          1985.0        20.0       White   
1  2005-01-18     08:15             M          1965.0        40.0       White   
2  2005-01-23     23:15             M          1972.0        33.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   
1      Speeding  Speeding             False         NaN     Citation   
2      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False



==== 남성인 데이터 ====


stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0      2005-01-02     01:55             M          1985.0        20.0   
1      2005-01-18     08:15             M          1965.0        40.0   
2      2005-01-23     23:15             M          1972.0        33.0   
3      2005-02-20     17:15             M          1986.0        19.0   
5      2005-03-23     09:45             M          1982.0        23.0   
...           ...       ...           ...             ...         ...   
91735  2015-12-31     20:25             M          1992.0        23.0   
91736  2015-12-31     20:27             M          1986.0        29.0   
91738  2015-12-31     20:45             M          1992.0        23.0   
91739  2015-12-31     21:42             M          1993.0        22.0   
91740  2015-12-31     22:46             M          1959.0        56.0   

      driver_race                   violation_raw         violation  \
0           White                        Speeding          Speeding   
1           White                        Speeding          Speeding   
2           White                        Speeding          Speeding   
3           White                Call for Service             Other   
5           Black  Equipment/Inspection Violation         Equipment   
...           ...                             ...               ...   
91735    Hispanic                        Speeding          Speeding   
91736       White                        Speeding          Speeding   
91738       White         Other Traffic Violation  Moving violation   
91739       White                        Speeding          Speeding   
91740    Hispanic                        Speeding          Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
5                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
91735             False         NaN       Citation       False      0-15 Min   
91736             False         NaN        Warning       False      0-15 Min   
91738             False         NaN        Warning       False      0-15 Min   
91739             False         NaN       Citation       False      0-15 Min   
91740             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
5                   False  
...                   ...  
91735               False  
91736               False  
91738               False  
91739               False  
91740               False  

[62895 rows x 14 columns]



==== 20세 데이터 ====


stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0      2005-01-02     01:55             M          1985.0        20.0   
43     2005-10-01     01:00             M          1985.0        20.0   
73     2005-10-02     06:10             M          1985.0        20.0   
125    2005-10-04     11:50             M          1985.0        20.0   
137    2005-10-04     23:10             M          1985.0        20.0   
...           ...       ...           ...             ...         ...   
91570  2015-12-21     20:22             F          1995.0        20.0   
91578  2015-12-22     03:18             M          1995.0        20.0   
91645  2015-12-27     13:50             F          1995.0        20.0   
91670  2015-12-28     10:38             F          1995.0        20.0   
91727  2015-12-31     09:17             F          1995.0        20.0   

      driver_race           violation_raw            violation  \
0           White                Speeding             Speeding   
43          White                Speeding             Speeding   
73          White                Speeding             Speeding   
125         White                Speeding             Speeding   
137         White                Speeding             Speeding   
...           ...                     ...                  ...   
91570       White                Speeding             Speeding   
91578    Hispanic                Speeding             Speeding   
91645       White  Registration Violation  Registration/plates   
91670       White                Speeding             Speeding   
91727       White                Speeding             Speeding   

       search_conducted search_type      stop_outcome is_arrested  \
0                 False         NaN          Citation       False   
43                False         NaN          Citation       False   
73                False         NaN          Citation       False   
125               False         NaN          Citation       False   
137               False         NaN  Arrest Passenger        True   
...                 ...         ...               ...         ...   
91570             False         NaN          Citation       False   
91578             False         NaN          Citation       False   
91645             False         NaN          Citation       False   
91670             False         NaN          Citation       False   
91727             False         NaN           Warning       False   

      stop_duration  drugs_related_stop  
0          0-15 Min               False  
43         0-15 Min               False  
73         0-15 Min               False  
125        0-15 Min               False  
137       16-30 Min               False  
...             ...                 ...  
91570      0-15 Min               False  
91578      0-15 Min               False  
91645      0-15 Min               False  
91670      0-15 Min               False  
91727     16-30 Min               False  

[3794 rows x 14 columns]

- 결과
  - 2005년 1월 데이터: 3개
  - 남성인 데이터: 62895개
  - 20세 데이터: 3794개

## 4) 데이터 연산
- 발생 이후 경과 일수
- 현재 - 발생일

In [54]:
print("==== stop_date의 Dtype 확인하기 ====\n")
police['stop_date'] = pd.to_datetime(police['stop_date'], yearfirst=True, exact=True)
print(police.info())

print("\n\n==== 상위 3개 확인하기 ====")
display(police.head(3))

==== stop_date의 Dtype 확인하기 ====

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   stop_date           91741 non-null  datetime64[ns]
 1   stop_time           91741 non-null  object        
 2   driver_gender       86406 non-null  object        
 3   driver_age_raw      86414 non-null  float64       
 4   driver_age          86120 non-null  float64       
 5   driver_race         86408 non-null  object        
 6   violation_raw       86408 non-null  object        
 7   violation           86408 non-null  object        
 8   search_conducted    91741 non-null  bool          
 9   search_type         3196 non-null   object        
 10  stop_outcome        86408 non-null  object        
 11  is_arrested         86408 non-null  object        
 12  stop_duration       86408 non-null  object        
 13  drugs_related

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02     01:55             M          1985.0        20.0       White   
1 2005-01-18     08:15             M          1965.0        40.0       White   
2 2005-01-23     23:15             M          1972.0        33.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   
1      Speeding  Speeding             False         NaN     Citation   
2      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False

In [57]:
police['days_since_event'] = pd.to_datetime("today")-police["stop_date"]
police.iloc[:, [0, -2, -1]]

stop_date  drugs_related_stop          days_since_event
0     2005-01-02               False 7484 days 13:23:42.740221
1     2005-01-18               False 7468 days 13:23:42.740221
2     2005-01-23               False 7463 days 13:23:42.740221
3     2005-02-20               False 7435 days 13:23:42.740221
4     2005-03-14               False 7413 days 13:23:42.740221
...          ...                 ...                       ...
91736 2015-12-31               False 3469 days 13:23:42.740221
91737 2015-12-31               False 3469 days 13:23:42.740221
91738 2015-12-31               False 3469 days 13:23:42.740221
91739 2015-12-31               False 3469 days 13:23:42.740221
91740 2015-12-31               False 3469 days 13:23:42.740221

[91741 rows x 3 columns]

In [58]:
police['days_since_event'].dt.days

0        7484
1        7468
2        7463
3        7435
4        7413
         ... 
91736    3469
91737    3469
91738    3469
91739    3469
91740    3469
Name: days_since_event, Length: 91741, dtype: int64